org.eclipse.xtext/src/org/eclipse/xtext/validation/INamesAreUniqueValidationHelper.java

```diff
@@ -8,29 +8,154 @@
  *******************************************************************************/
 package org.eclipse.xtext.validation;
 
+import java.util.Collection;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
 import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.ISelectable;
+import org.eclipse.xtext.resource.IResourceDescription.Delta;
+import org.eclipse.xtext.resource.impl.IsAffectedExtension;
+import org.eclipse.xtext.scoping.ICaseInsensitivityHelper;
 import org.eclipse.xtext.util.CancelIndicator;
 
 import com.google.inject.ImplementedBy;
+import com.google.inject.name.Named;
 
 /**
+ * The helper can be used to validate unique names locally or across resource boundaries.
+ * 
+ * @see LocalUniqueNameContext
  * @author Sebastian Zarnekow - Initial contribution and API
  */
 @ImplementedBy(NamesAreUniqueValidationHelper.class)
 public interface INamesAreUniqueValidationHelper {
 
 	/**
-	 * Create errors for objects that have the same name. Objects, that do not belong to
+	 * Provide the context information for the unique name validation.
+	 * 
+	 * Registered by default as a {@link IsAffectedExtension} as {@link Named @Named("IsAffectedExtension.UniqueNames")}.
+	 * 
+	 * @since 2.22
+	 */
+	@ImplementedBy(DefaultUniqueNameContext.ExportedFromResource.class)
+	interface ContextProvider extends IsAffectedExtension {
+		
+		/**
+		 * Obtain the context of the unique name validation for the given resource.
+		 * 
+		 * May return null.
+		 */
+		Context tryGetContext(Resource resource, CancelIndicator cancelIndicator);
+		
+		@Override
+		default boolean isAffected(Collection<Delta> deltas, IResourceDescription candidate,
+				IResourceDescriptions context) {
+			return false;
+		}
+	}
+	
+	/**
+	 * <p>
+	 * Provides context information to the validation that allows to introspect the scope in which the validation for
+	 * unique names is supposed to happen.
+	 * </p>
+	 * <p>
+	 * The term <code>clusterType</code> is used to denote a {@link EClass (super-)type} of the current object under
+	 * validation. When checking the name for uniqueness, all objects that are instances of the given cluster type are
+	 * considered. This can be used to define different namespaces. For example, a language that has the concepts of
+	 * <code>Fields</code>, <code>Procedures</code> and <code>Functions</code> may provide 2 different cluster types.
+	 * <ol>
+	 * <li>All fields must have a unique name.</li>
+	 * <li>The executable procedure and functions use the same namespace and may not have duplicate names.</li>
+	 * </ol>
+	 * This would allow a field to use the same name as a procedure but flag a duplication if a procedure has the same
```

Comment:
`procedure` -> `procedure/function` to indicate that a field can have the same name as a procedure or function